Factory-create the elliptic background and initial guess, reduce number of elliptic executables

nilsvu opened the pull request:
## Proposed changes

This is a maintenance PR for the elliptic executables. It separates the concepts of the background, boundary conditions and the initial guess for elliptic systems, so I can work on them and upgrade them separately. Here's some details and plans for the future for each of these:

- With _background_ I mean the variable-independent fields in the equations, e.g. the fixed-sources f(x) in a Poisson equation -Laplace(u)=f(x), the matter-density rho(x) in a TOV-solve or the conformal metric in an XCTS solve. Currently this is always given by an analytic solution, but I will of course add analytic data that are not analytic solutions, e.g. the conformal metric for black-hole binaries in various coordinates. My plan is to make these factory-creatable so one can specify e.g. SKS or SHK initial data in the input file. **Edit: Done, and this PR switches the executables over to the factory-creation.**
- The _boundary conditions_ are currently given by the analytic solution as Dirichlet conditions, but they will use the boundary-condition mechanism once it's implemented in the domain. At that point they will also be made factory-creatable so they'll be specified in the input file. An example is to use apparent-horizon b.c. on some boundaries and conformal flatness on others. **Edit: Done, but the current executables don't use the factory-created b.c. yet. The upcoming elliptic DG operator will switch over to those.**
- The _initial guess_ is currently always zero, but it will support any analytic solution or analytic data, also specifiable in the input file once those are factory-creatable. In the next PR I'll add support for non-zero initial guesses (I omitted it from this PR to keep it smaller). **Edit: Factory-creation is done (see "background"). Non-zero initial guesses coming in separate PR.**

### Upgrade instructions

<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add elliptic background and inital guess tags

nilsvu committed:
Factory-create the elliptic background and initial guess

- Define an elliptic problem by its "background" (the variable-independent fields
  in the equations), its boundary conditions and its initial guess. Separate these
  concepts in the executables.
- The background and the initial guess can now be factory-created from options
  in the input file. This reduces the number of executables and leaves only
  'SolvePoisson{1,2,3}D' and 'SolveElasticity{2,3}D' for now.
- Factory-creatable boundary conditions will be supported with the upcoming elliptic
  DG operator.

moxcodes commented on src/Elliptic/Protocols.hpp:
```diff
@@ -0,0 +1,21 @@
+
+struct AnalyticSolution {
+  template <typename ConformingType>
+  struct test {
```
Why not make it a base class immediately? My rough understanding of protocols is that they augment base classes with the ability to insist on other compile-time interface features, but if you're not going to use those, it seems better to stick with the standard C++ inheritance features.

nilsvu replied:
You're probably right, I meant to do the whole inheritance and factory-creation thing later but I might as well do it now :)

nilsdeppe commented on src/Elliptic/Tags.hpp:
```diff
@@ -37,5 +66,23 @@ struct FluxesComputer : db::SimpleTag {
 };
 
+template <typename BackgroundType>
+struct Background : db::SimpleTag {
```
Please add dox

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/Poisson/Zero.hpp:
```diff
@@ -0,0 +1,75 @@
+/// The trivial solution \f$u=0\f$ of a Poisson equation. Useful as initial
+/// guess.
+template <size_t Dim>
+class Zero : public tt::ConformsTo<elliptic::protocols::AnalyticSolution> {
```
Since the implementation is basically the same for all systems, can you do something similar to what we did with `PointwiseFunctions/AnalyticSolutions/Hydro/SmoothFlow.?pp`, where there's a generic implementation and then the systems just specialize what they need to? In this case it seems to be just `supported_tags` that I think is different?

nilsvu replied:
Are you suggesting to share this implementation between the elliptic systems? For Poisson and Elasticity this could work I think, since the trivial solution is to set all fields to zero. For XCTS the trivial solution involves some fields are 1 (like the conformal factor) and I would call this `Flat` instead of `Zero`. Do you still think I should combine the classes?

nilsdeppe replied:
That sound similar to what we do with hydro still, yea. Basically, we add extra functions to set the variables that either aren't defined or are defined incorrectly in the base/shared class

nilsvu replied:
Hm I'm actually not sure the base class would make the code shorter or easier to understand. The subclasses would need to pull in constructors from the base, forward the variables function and still define equality operators.. Though I agree with you that it would be good to derive from a base class if any non-trivial logic could be shared, I think here it is not worth the effort and would actually make the code longer and more complicated.

nilsdeppe replied:
Sorry, how would the 1 line for constructors and 1 line for `variables` be longer?

```
using shared_elliptic::options;
using shared_elliptic::shared_elliptic;
using shared_elliptic::variables;
using shared_elliptic::pup;
```
that seems like the "long" implementation here

nilsvu replied:
Won't I need to implement `variables` and forward, because it's a template, and also define equality operators?

nilsvu replied:
I'll also need tests for both derived classes _and_ the base class

nilsdeppe replied:
No, I would template the base class on the `supported_tags` and then you can use the exact implementation you have here, I think. 

Yes, you need tests for both.

If you feel that strongly that it would make the code harder to maintain, you can leave it :)

Edit:
I would advocate that for less trivial solutions that are shared (if there are any) that you consider sharing them, but I can see how this case might be too simple to make it worth the effort.

nilsleiffischer commented:
Pushed a fixup. Cheers 👍

nilsvu commented:
Pushed a fixup. Cheers 👍

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash

moxcodes commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -89,16 +89,17 @@ struct InitializeSystem {
                   num_grid_points, 0.};
```
Commit message is too long, splits the line in github (probably command line too, but I didn't check)

tomwlodarczyk commented on src/PointwiseFunctions/AnalyticSolutions/Poisson/Zero.hpp:
```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <cstddef>
+#include <limits>
```
What do you need the limits for here?

nilsvu replied:
I don't, removed 👍

tomwlodarczyk commented on src/PointwiseFunctions/AnalyticSolutions/Poisson/Zero.hpp:
```diff
@@ -0,0 +1,75 @@
+      tmpl::list<Tags...> /*meta*/) const noexcept {
+    static_assert(tmpl::size<tmpl::list_difference<tmpl::list<Tags...>,
+                                                   supported_tags>>::value == 0,
+                  "The requested tag is not supported");
```
As I understand it, you have to call `variables()` with all `supported_tags`? Are there instances where only one `Tag` is required or is it something to worry about?

tomwlodarczyk replied:
Ah sorry, it seems `list_difference` returns A \ B, rather than (A \ B) U (B \ A)

nilsvu replied:
As you say, the `variables` function can be called with any subset of the supported tags

nilsvu commented:
@moxcodes in response to your comment I did the work and added the abstract base classes to avoid adding the protocol as a temporary solution. I pushed the changes as a series of fixup commits to make them (hopefully) easier to review. Once you're happy with the changes I'll rebase and squash everything down to a consistent git history. Could you and @nilsdeppe take a look at the fixups please?

If you prefer I can factor some of the changes out into a separate PR.

The result of the changes is this:
- The evolution executables are unchanged, though `ObserveFields` and `ObserveErrorNorms` now support checking at runtime whether or not an analytic solution is available. The check is disabled when no analytic solution tensors are given at compile-time.
- Only two elliptic executables remain (modulo dimensions): `SolvePoisson{1,2,3}D` and `SolveElasticity{2,3}D`. The background and initial guess can be selected in the input file; boundary conditions will follow once the domain creators support them.
- The elliptic analytic solutions are implemented way cleaner now, IMO. A lot of boilerplate code related to template-specializing and instantiating `variables` functions is gone and replaced by `CachedTempBuffer`s that Will introduced for the Kerr solution. That also encourages caching and re-using quantities in the computations, though I haven't refactored the computations to take advantage of that yet.

moxcodes commented:
@nilsleiffischer I think the fixups are so numerous and alter code already altered in the existing commits to the extent that I'd just prefer to re-examine the squashed collection of commits. Once @nilsdeppe gives the go-ahead, I'd say go ahead and squash, and I'll look at it again. Thanks!

nilsdeppe commented:
yea... this basically seems to rewritten the entire PR...

nilsvu commented:
@moxcodes @nilsdeppe OK, it's done!

moxcodes commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -32,47 +32,53 @@ namespace Actions {
  */
-template <typename AnalyticSolutionTag, typename AnalyticSolutionFields>
+template <typename BackgroundTag, typename AnalyticSolutionFields,
+          typename AnalyticSolutionType = typename BackgroundTag::type>
```
I'm not sure I understand the type logic here -- the documentation makes me think that the code is going to check that the quantity in `BackgroundTag` is an analytic solution according to a globally agreed-upon type that determines an analytic solution, but whenever the calling code doesn't specify the `AnalyticSolutionType`, the default template parameter will ensure that it will be treated as an analytic solution.

nilsvu replied:
In that case the code won't actually compile if the type isn't an analytic solution, because the compiler tries to instantiate its `variables` function and fails if it doesn't provide the required tags. And if the class provides the tags then it might as well be treated as an analytic solution :)

moxcodes replied:
Okay, but I thought that kind of 'loose' interface specification (where it doesn't matter the inheritance path, as long as it defines the right functions it 'works') was precisely what you were attempting to avoid by specifying base classes / protocols?

nilsvu replied:
Yea in fact the type passed as background here might not actually be an analytic solution, even if it provides the required tags - it could just be an initial guess with data for all fields, but that doesn't necessarily solve the equations. I'll have to remove the default parameter in a separate commit though or squash the two commits together with the one upgrading the executables, because else the commit won't compile anymore.

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/AnalyticSolution.hpp:
```diff
@@ -0,0 +1,61 @@
+        tuples::TaggedTuple<RequestedTags...>,
+        tmpl::filter<
+            creatable_classes,
+            std::is_base_of<AnalyticSolution<Dim, Registrars>, tmpl::_1>>>(
```
Is the only difference between this and the function it inherits from `AnalyticData.hpp` that it filters the `creatable_classes` for types that inherit from the current type? Can it just be a compile-time check that all of the `creatable_classes` inherit from the current type? I'm not sure I see a case where where we'd want the `creatable_classes` type alias in a particular base class to contain classes that do not inherit from that base class.

nilsvu replied:
1. Yes
2. No, because only a subset of the list of registrars are analytic solutions, and only for those can the `variables` function template be instantiated for these tags. On other words: The user can specify any analytic data or solution in the input file, but the analytic solution tags can only be retrieved from a subset of the creatable classes.

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Poisson/AnalyticSolution.hpp:
```diff
@@ -0,0 +1,59 @@
+        tuples::TaggedTuple<RequestedTags...>,
+        tmpl::filter<
+            creatable_classes,
+            std::is_base_of<AnalyticSolution<Dim, Registrars>, tmpl::_1>>>(
```
Same question as for the Elasticity base class.

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp:
```diff
@@ -4,26 +4,60 @@
+
+namespace detail {
+template <typename DataType>
+struct BentBeamVariables {
```
This detail class seems unnecessary -- why use this instead of just creating template specializations of the `variables` function? I would have thought that supporting the retrieval of solutions via the `variables` function was the entire point of the `dynamic_cast` trickery in the base class.

nilsvu replied:
The issue is the template on registrars. That prevents moving the implementations to the cpp file. If you have an idea how to deal with that I'd be happy to hear 😄

moxcodes replied:
Can you clarify why you need to use the registrars trick here? I would have thought that, for instance, for the Poisson analytic solutions, you could always just explicitly specify all of the Poisson analytic solutions in its `creatable_classes`?

nilsvu replied:
Hm, that would mean the executable can't choose which analytic solution or data classes it allows, but it's always the full list of available classes. I'll have to think if that can work in the future. For more involved systems like XCTS I'll likely have more than one "solve this system" executable because they need to do more things, e.g. a neutron-star solve needs to interpolate data to the center of the star in every nonlinear step and a vacuum solve may want to find apparent horizons instead. This might be solvable with more base classes.. I'll think about it a bit more :)

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Elasticity/HalfSpaceMirror.cpp:
```diff
@@ -183,36 +181,26 @@ tuples::TaggedTuple<Tags::Strain<3>> HalfSpaceMirror::variables(
-    tmpl::list<::Tags::FixedSource<Tags::Displacement<3>>> /*meta*/) noexcept {
-  return {make_with_value<tnsr::I<DataVector, 3>>(x, 0.)};
+template <typename DataType>
+void HalfSpaceMirrorVariables<DataType>::operator()(
```
Same question about the replacement of variables specialization with operator() in a detail class design you've chosen

moxcodes commented on src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp:
```diff
@@ -4,68 +4,59 @@
-    tmpl::list<Tags::Field> /*meta*/) noexcept {
-  return {Scalar<DataVector>(1. / sqrt(1. + get(dot_product(x, x))))};
+template <typename DataType, size_t Dim>
+void LorentzianVariables<DataType, Dim>::operator()(
```
And here - same design change from `variables` specializations to `operator()`s in a detail class

nilsdeppe commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -32,47 +32,53 @@ namespace Actions {
 /*!
```
This change has nothing to do with the rest of the PR and needs to be factored out. (so do many of the other commits as far as I can tell)

nilsdeppe commented on src/Elliptic/Executables/Elasticity/CMakeLists.txt:
```diff
@@ -35,8 +35,8 @@ endfunction(add_elasticity_executable)
 add_elasticity_executable(
```
This also should be a separate PR.

nilsdeppe commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -26,15 +26,12 @@ namespace Actions {
  *
```
These `Upgrade elliptic executables` type short git messages aren't that helpful. This seems to be doing at least 3 things at once. Can they not be separated out?

nilsdeppe reviewed (changes requested):
When a PR starts exploding like this and making unrelated changes or changes that one could never guess at from the title the PR needs to be factored out. As is, this is sort of a (unintentional) "let's see what can all be snuck by people who won't bother looking because the name is not what the PR is doing."

As far as I can tell only one, maybe 2 of the commits in this PR are actually related to the title.

nilsvu commented:
Sure, I can factor this out in separate PRs. I have become hesitant doing that right away because I have made the experience that I get comments like "why are you making this change for no apparent reason, don't do it, let's close this PR" because the context is missing. That's why I open the full-context PR first and split into smaller ones when reviewers ask for it.

nilsvu committed:
Fix penalty parameter in an input file

The penalty parameter should be (close to) 1.

nilsvu commented:
@nilsdeppe and @moxcodes The dependencies for this PR are now merged, so it has shrunken to just switching over the elliptic executables to the factory-creation mechanism. Essentially it reduces the number of elliptic executables from one per analytic solution / data down to `SolvePoisson{1,2,3}D` and `SolveElasticity{2,3}D`. Please take another look when you get the chance!

A few notes on future plans:
- I might in the future add specialized executables, like a `SolveThermalNoise` executable, which allow only particular choices for the background, initial guess and boundary conditions and which may do additional problem-specific things.
- I haven't quite decided yet how many flavors of XCTS(+hydro) executables I'll end up with; some notes on that are written up in https://github.com/sxs-collaboration/spectre/issues/2695. Those that do algorithmically different things, like finding apparent horizons or not, will certainly be different executables. Also, I might have separate executables that are optimized for conformal-flatness or solve only a subset of the equations.

nilsvu commented:
Wow that was fast, thanks for reviewing @nilsdeppe and @moxcodes!
